Desktop_Interface/isobuffer.cpp

```diff
@@ -2,391 +2,364 @@
 {
-    //ignore singleBit for now
-    double timeBetweenSamples = (double) sampleWindow * (double) samplesPerSecond / (double) numSamples;
-    double accumulatedDelay = 0;
-    int delaySamples = (int)((double)delayOffset * (double)samplesPerSecond);
-
-    int front = back - 1 - delaySamples;
-    if (front < 0) front = 0;
-    int idx, subIdx;
-    if(readData!=NULL) free(readData);
-    readData = (short *) calloc(numSamples, sizeof(short));
-
-    if(singleBit){
-        for (int i=0; i<numSamples;i++){
-            if (timeBetweenSamples > (double) front){
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-            }
-
-            idx = (int) floor(((double) front - accumulatedDelay));
-            subIdx = (int) floor(8*(((double) front - accumulatedDelay) - floor(((double) front - accumulatedDelay))));
-
-            //qDebug() << "subIdx = " << subIdx;
-
-            if (idx < 0){
-                accumulatedDelay--;
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-                idx = (int) round(((double) front - accumulatedDelay));
-            }
-            readData[i] = buffer[idx] & (1 << subIdx);
-
-            accumulatedDelay += timeBetweenSamples;
-        }
-    }else{
-        for (int i=0; i<numSamples;i++){
-            if (timeBetweenSamples > (double) front){
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-            }
-
-            idx = (int) round(((double) front - accumulatedDelay));
-            if (idx < 0){
-                accumulatedDelay--;
-                accumulatedDelay -= (double) front;
-                front = bufferEnd;
-                idx = (int) round(((double) front - accumulatedDelay));
-            }
-            readData[i] = buffer[idx];
-
-            accumulatedDelay += timeBetweenSamples;
-        }
-    }
-
-    return readData;
+	/* Refactor Note:
+	 *
+	 * Refactoring this function took a few passes were i made some assumptions:
+	 *  - round() should be replaced by floor() where it was used
+	 *  - int(floor(x)) and int(x) are equivalent (since we are always positive)
+	 *  - free(NULL) is a no-op. This is mandated by the C standard, and virtually all
+	 * implementations comply. A few known exceptions are:
+	 *    - PalmOS
+	 *    - 3BSD
+	 *    - UNIX 7
+	 *   I do not know of any non-compliant somewhat modern implementations.
+	 *
+	 * The expected behavior is to cycle backwards over the buffer with a stride of
+	 * timeBetweenSamples steps, and insert the touched elements into readBuffer.
+	 *
+	 * ~Sebastian Mestre
+	 */
+	const double timeBetweenSamples = sampleWindow * samplesPerSecond / numSamples;
+	const int delaySamples = delayOffset * samplesPerSecond;
+
+	free(readData);
+	readData = (short*) calloc(numSamples, sizeof(short));
+
+	double itr = delaySamples + 1;
+	for (int i = 0; i < numSamples; i++)
+	{
+		while (itr > insertedCount)
```

Comment:
I'm not quite sure that this is the best behaviour.  It looks like it's providing more recent data for the case when the delay is larger than the valid buffer length.  In this case, we probably just want to return nullptr (and check for errors upstream).

Reply:
Oh, I see. I'll fix that then. I might have some questions later, about what would be a good way to handle nulls upstream, if that's ok.